### ISSUE TITLE
Investigate instability of llama models post commit test in main

### ISSUE THREAD

**Issue by tt-rkim** [tenstorrent/tt-metal#14474]
due to commit https://github.com/tenstorrent/tt-metal/commit/c46e5019d95a179f1446f216e0518baf70410536

cc: @TT-billteng @bkeith-TT @uaydonat @yieldthought

**Comment by uaydonat** [comment]
are the failures due to pcc or hangs?

@caixunshiren let's investigate... but keep in mind #9370 is still open, they only pushed a work-around.

**Comment by caixunshiren** [comment]
> are the failures due to pcc or hangs?
> 
> 
> 
> @caixunshiren let's investigate... but keep in mind #9370 is still open, they only pushed a work-around.

Still investigating. Not able to reproduce locally. Shouldn't be a hang or ndpcc because it passed the flash decode ndpcc tests 🤔

**Comment by uaydonat** [comment]
We cannot disable llama, it is our most important model.

We might rather need to revert @xuncaiTT 's commit if it is the culprit.

It has been one day, if we do not have a solution, let's revert the optimization, work on it offline.

**Comment by caixunshiren** [comment]
@cglagovichTT was able to repro a seg fault on llama stress test on a commit before my changes. It is likely that this failure is not related to my commit.

**Comment by TT-billteng** [comment]
@tt-rkim what is the exact test this issue is about?

> @cglagovichTT was able to repro a seg fault on llama stress test on a commit before my changes. It is likely that this failure is not related to my commit.

So where did this issue first appear? Did we git bisect? Do we have an exact command to repro the error, even if it's ND?

**Comment by caixunshiren** [comment]
Yes it is the exact same test. It was done using git bisect. @cglagovichTT would you be able to add more details?

**Comment by cglagovichTT** [comment]
Is this a duplicate of https://github.com/tenstorrent/tt-metal/issues/14475?

My repro command is:
```
FAKE_DEVICE=N300 python -m  pytest --count=60  -svv models/demos/llama3/demo/demo.py::test_llama_demo -k "instruct_weights-32"
```
The command runs for 1 hour. When the test fails, it might be after 20 minutes or 45 minutes of execution.

My most recent bisect points me to this commit as the first bad.
https://github.com/tenstorrent/tt-metal/commit/0af26ed
I must have had a false negative during my bisect since this commit only changes test code. I'll start another bisect and run tests for 2 hours before determining a commit is good.

Note that it's possible that this segfault has existed in one way or another for many weeks, just becoming more or less prevalent as commits change the timing of threads.

**Comment by tt-rkim** [comment]
We can combine the two if it's the same issue

Also that's a crazy commit to be causing issues... unless you mean the chunk of changes because @mrshaw01 didn't squash.

**Comment by mrshaw01** [comment]
Is that related. I mean, how would changes in the unit tests of operations cause post-commit tests for llama models to fail.

FYI, details of the PR: https://github.com/tenstorrent/tt-metal/pull/13849/.

**Comment by cglagovichTT** [comment]
We thought it was solved by https://github.com/tenstorrent/tt-metal/pull/14273
@mtairum are you able to repro this issue locally? 
https://github.com/tenstorrent/tt-metal/actions/runs/11720766141/job/32647229455

**Comment by tt-rkim** [comment]
Have you tried getting access to a VM? You are able to provision one yourself.

**Comment by ttmchiou** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/11728678704/job/32673138552
https://github.com/tenstorrent/tt-metal/actions/runs/11729309675/job/32679768714
https://github.com/tenstorrent/tt-metal/actions/runs/11729454024/job/32675633969
https://github.com/tenstorrent/tt-metal/actions/runs/11728678704/job/32673138552
https://github.com/tenstorrent/tt-metal/actions/runs/11727420471/job/32668842127 (N150 + N300 failure here)
https://github.com/tenstorrent/tt-metal/actions/runs/11727271604/job/32668348645
^Failing tests 

seems to be ND and only on WH cards.

**Comment by mtairum** [comment]
So, i've created this branch https://github.com/tenstorrent/tt-metal/tree/refs/heads/mtairum/debug_llama_post_commit where I'm running on N300 (prior I've tested on N150 as well), just the Llama3 tests: 1B / 3B / 8B.

Like @ttmchiou mentioned, these are ND and they occur on N150 and N300.
example of a N300 failure here from my branch: https://github.com/tenstorrent/tt-metal/actions/runs/11741524983/job/32710638499

The failures do occur often enough, if you look at the post-commit run history for my branch (4 out of 11 failed on the llama test, 1 failed on post-run):
https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch%3Amtairum%2Fdebug_llama_post_commit

I've been trying to reproduce this locally on the VM `172.27.45.28`, a  `tt-metal-dev-common-whx2-1` and have run so far more than 10 times the Llama3 1B/3B/8B test loop and never got a failure.
I should've got a failure by now.

Do you think it could be something related to specific runners?

The way I'm building tt-metal on the VM is through the `./build_metal.sh` script, and then I run the `/tests/scripts/run_python_model_tests.sh` directly.

I see from the github scripts that the model tests are run from TT-Metal provided by wheels, and not directly the source. Could this be a problem?

**Comment by dimitri-tenstorrent** [comment]
@mtairum Hi! 
I can not directly see how the wheel could affect this but also the failure is not very clear
```
worker 'gw0' crashed while running 'models/demos/llama3/tests/test_llama_model.py::test_llama_model_inference[wormhole_b0-True-2-quick]'
```

One way to solve this would be to test your runs on the lab machine that uses the wheel. 
1. Remove the source code installation so that the machine is clean to make sure we can run a good test and identify if the wheel is a problem.
2. Download the wheels from one of your successful runs like from here: 
https://github.com/tenstorrent/tt-metal/actions/runs/11742511348
[eager-dist-ubuntu-20.04-wormhole_b0](https://github.com/tenstorrent/tt-metal/actions/runs/11742511348/artifacts/2163058986)
3. Install it with `pip install $wheel_filename`

Execute your tests and see if you can get the `worker crashed` failures based of using the wheel.

**Comment by mrshaw01** [comment]
FYI, when the test `models-unit-tests (wormhole_b0, N150) / model wormhole_b0 N150` failed, I clicked on "run only failed tests" seven times, but it still failed. However, when I reran the entire "all post commit tests," it passed.

The job: https://github.com/tenstorrent/tt-metal/actions/runs/11758161892/job/32758797445
Hope it helps.

**Comment by jvasilje** [comment]
@uaydonat what's the plan for this P0 bug? It has been open for 2 weeks.

**Comment by mtairum** [comment]
I've tried the wheel, no luck in replicating the issue on lab machine.

We do have a few in flight small changes that could provoke a test failure: testing those now on CI to see if they pass this issue.

**Comment by tt-rkim** [comment]
I also recommend trying custom test dispatch: https://github.com/tenstorrent/tt-metal/actions/workflows/test-dispatch.yaml

as that should replicate an environment close to CI in post commit.

Wheels likely won't be a problem, but you could try installing the wheel locally to see if it helps track down the issue.

**Comment by tt-rkim** [comment]
> take out a CI machine temporarily for interactive testing to see if it is a machine dependent issue.

We can target a particular machine you've seen with custom test dispatch as I mentioned above. Please let us know if there's anything blocking from doing this.

> lower clock to 900MHz to see if the hang goes away
> turn back on di/dt work-arounds to see if the hangs go away

I believe your team has the necessary tools to do those without being blocked.

**Comment by uaydonat** [comment]
It is not blocking progress on other models, just the llama new codebase delivery. 

Lowering to P1, but it is still important and @mtairum is working on this full time.

**Comment by tt-rkim** [comment]
Thanks @uaydonat 
Yes this sounds like P1, thanks @mtairum for taking a look
Pls let us know if you need any more resources

**Comment by TT-billteng** [comment]
Can we disable? This can't be the ONLY llama test in post-commit I hope?

**Comment by uaydonat** [comment]
@mtairum let's disable the ND tests, they are not helping if they fail anyways.

**Comment by mtairum** [comment]
I've disabled the tests for now to avoid the ND fails on the pipeline.
PR Merged here: https://github.com/tenstorrent/tt-metal/pull/14968

Recent small fixes to the model added today didn't change the behaviour.

My current plan is to identify a pattern on CI VMs that show consistent hangs, if any. 
Then I'll try to reproduce locally on one of those following the same build process as CI.

**Comment by TT-billteng** [comment]
> I've disabled the tests for now to avoid the ND fails on the pipeline. PR Merged here: [#14968](https://github.com/tenstorrent/tt-metal/pull/14968)
> 
> Recent small fixes to the model added today didn't change the behaviour.
> 
> My current plan is to identify a pattern on CI VMs that show consistent hangs, if any. Then I'll try to reproduce locally on one of those following the same build process as CI.

Thanks!

**Comment by esmalTT** [comment]
@mtairum Have you tried to run the test with watcher? Even if you run it on a machine where you're not seeing the hang, it might catch an illegal read/write that might help you isolate the issue.

**Comment by uaydonat** [comment]
@mtairum any updates?

**Comment by mtairum** [comment]
Yes.

Tried to run on CI with watcher enabled, didn't get any extra information.

However I've compiled a list of passing/failing machines based on 44 runs. (N150 only). From all the CI VMs that I got multiple times, they either always pass or always fail.

@ttmchiou @tt-rkim @TT-billteng is it possible to get access to `tt-metal-ci-vm-144` to debug the llama failing? That machine consistently fails(4 out of 4 times).

| CI VM | Pass | Fail |
|------------------|------|------|
| tt-metal-ci-vm-27| 1    | 0    |
| tt-metal-ci-vm-29| 1    | 0    |
| tt-metal-ci-vm-31| 2    | 0    |
| tt-metal-ci-vm-32| 1    | 0    |
| tt-metal-ci-vm-57| 0    | 1    |
| tt-metal-ci-vm-58| 4    | 0    |
| tt-metal-ci-vm-68| 2    | 0    |
| tt-metal-ci-vm-95| 1    | 0    |
| tt-metal-ci-vm-98| 1    | 0    |
| tt-metal-ci-vm-120| 0   | 2    |
| tt-metal-ci-vm-124| 0   | 1    |
| tt-metal-ci-vm-126| 1   | 0    |
| tt-metal-ci-vm-127| 0   | 1    |
| tt-metal-ci-vm-130| 0   | 1    |
| tt-metal-ci-vm-133| 0   | 1    |
| tt-metal-ci-vm-134| 0   | 3    |
| tt-metal-ci-vm-137| 0   | 3    |
| tt-metal-ci-vm-138| 0   | 1    |
| tt-metal-ci-vm-139| 0   | 1    |
| tt-metal-ci-vm-140| 0   | 1    |
| tt-metal-ci-vm-143| 0   | 1    |
| tt-metal-ci-vm-144| 0   | 4    |
| tt-metal-ci-vm-146| 6   | 0    |
| tt-metal-ci-vm-149| 3   | 0    |

**Comment by uaydonat** [comment]
This table is very interesting. It appears that the problem is machine specific.

Yes, @tt-rkim let's get access to 144, try all the di/dt workarounds, collect more data, then we will probably need to involve milos and syseng.

**Comment by ttmchiou** [comment]
@tt-rkim is on vacation.
I can handle getting @mtairum access to 144.
Will DM privately to set up access

**Comment by ttmchiou** [comment]
special note,
VM-144 seems to only have 4 cores as noticed by @TT-billteng.

**Comment by mtairum** [comment]
Small update:

Initial testing on 144 shows some things:

- 1B and 8B tests both seem work fine
- 3B fails

Running with strace I see the following:
```
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_KILLED, si_pid=3840555, si_uid=1000, si_status=SIGKILL, si_utime=1209, si_stime=944} ---
```

After asking chat GPT a bit more it seems that the `SIGKILL` can be due to OoO 🤔 
This is weird since the 8B weights are larger than 3B...

**Comment by mtairum** [comment]
Found the root cause of the OoO issue 🙌 

When loading the transformer, in 3B specifically, we were going over the 20GB of allocated memory, hence why some VMs constantly crash, while others don't.


The hint was that 1B and 8B were passing but 3B was not.

The MVP here was the command:
```
import psutil, os ; process = psutil.Process(os.getpid()) ; print(f"Memory usage: {process.memory_info().rss / 1024**2:.2f} MB")
```
which told me where in the code the memory was suddenly exploding in 3B.


In reality, the issue was both in 1B and 3B, but since 1B is small enough it was still under the 20GB VM limit. After further investigation, the culprit was the max sequence length.

By default these models go up to 128K context length. In practice, we'd only want to use the maximum length for demo. For our small tests we don't really want to test out such large sequences: 128 sequence length is more than ok. 
However, when instantiation our model, we had an override specifically set for 8B and 70B on single device: reduce the max seqlen from 128k to 32K. This was an artifact from older code that we didn't notice when bringing up the unified codebase.

When generating the cos/sin for the rotary matrices, these take the sequence length in use: So for 8B they would be pegged to 32K while in 1B/3B would be 128k! 1B was not failing because it was just shy of the maximum 20GB of memory on the CI VM: it was just lucky.

Coincidentally, this test behaviour is actually something I started to address in a different issue/PR (in branch `mtairum/paged_attn_llama3`)

For the sake of this ticket I've pushed a quick patch to limit the max_seq_len and also reduces the max_seq_len even further when running a 1-layer llama model - i.e., a test or debug run, to speed CI up. PR here: https://github.com/tenstorrent/tt-metal/pull/15111

This quick patch should support both N150 and N300, so I've re-enabled the tests in the pipeline. This also explains why we were previously running out of memory for 11B: it wasn't reducing the sequence length, thus going OoO 😢.  Also enabled that one.

After all the CI pipelines pass I'll close this ticket. VM-144 should be good to go back to CI, but just in case, let's leave it out until Monday in case something doesn't work as expected. I'll ping @ttmchiou on this.


Additionally, I've created another P2 issue to further alleviate CI after discussion with @yieldthought: https://github.com/tenstorrent/tt-metal/issues/15114  (FYI: @uaydonat 

In short, to avoid spending time loading the TT model and the reference model, on import model families like Llama3 we should include reference output tensors; if these are present, instead of loading the reference model and run it, we can just compare the TT model output against a previously cached reference. This will dramatically step up CI, but some minutes for every job that includes Llama.

**Comment by mtairum** [comment]
Pushed the fix to main. 

I've tested 6 runs of all-post-commit and got one of the previous OoO machines which is now passing 👍 

Going to monitor for the following days in case any other OoO memory issue appears that my quick patch didn't cover, as it was machine dependent.
If needed I'll reopen this issue.

**Comment by mtairum** [comment]
@ttmchiou you can set VM-144 back to CI 🙏  
Thanks for the access.
